### PR TITLE
[SPARK-34506][CORE] ADD JAR with ivy coordinates should be compatible with Hive transitive behavior

### PR DESCRIPTION
**Pull request by shardulm94**
### What changes were proposed in this pull request?
SPARK-33084 added the ability to use ivy coordinates with `SparkContext.addJar`. PR #29966 claims to mimic Hive behavior although I found a few cases where it doesn't

1) The default value of the transitive parameter is false, both in case of parameter not being specified in coordinate or parameter value being invalid. The Hive behavior is that transitive is [true if not specified](https://github.com/apache/hive/blob/cb2ac3dcc6af276c6f64ee00f034f082fe75222b/ql/src/java/org/apache/hadoop/hive/ql/util/DependencyResolver.java#L169) in the coordinate and [false for invalid values](https://github.com/apache/hive/blob/cb2ac3dcc6af276c6f64ee00f034f082fe75222b/ql/src/java/org/apache/hadoop/hive/ql/util/DependencyResolver.java#L124). Also, regardless of Hive, I think a default of true for the transitive parameter also matches [ivy's own defaults](https://ant.apache.org/ivy/history/2.5.0/ivyfile/dependency.html#_attributes).

2) The parameter value for transitive parameter is regarded as case-sensitive [based on the understanding](https://github.com/apache/spark/pull/29966#discussion_r547752259) that Hive behavior is case-sensitive. However, this is not correct, Hive [treats the parameter value case-insensitively](https://github.com/apache/hive/blob/cb2ac3dcc6af276c6f64ee00f034f082fe75222b/ql/src/java/org/apache/hadoop/hive/ql/util/DependencyResolver.java#L122).

I propose that we be compatible with Hive for these behaviors

### Why are the changes needed?
To make `ADD JAR` with ivy coordinates compatible with Hive's transitive behavior

### Does this PR introduce _any_ user-facing change?

The user-facing changes here are within master as the feature introduced in SPARK-33084 has not been released yet
1. Previously an ivy coordinate without `transitive` parameter specified did not resolve transitive dependency, now it does.
2. Previously an `transitive` parameter value was treated case-sensitively. e.g. `transitive=TRUE` would be treated as false as it did not match exactly `true`. Now it will be treated case-insensitively.

### How was this patch tested?

Modified existing unit tests to test new behavior
Add new unit test to cover usage of `exclude` with unspecified `transitive`

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @AngersZhuuuu @wangyum FYI

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -1224,7 +1224,10 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
   test("SPARK-33084: Add jar support Ivy URI in SQL") {
     val testData = TestHive.getHiveFile("data/files/sample.json").toURI
     withTable("t") {
-      sql(s"ADD JAR ivy://org.apache.hive.hcatalog:hive-hcatalog-core:$hiveVersion")
+      // exclude org.pentaho:pentaho-aggdesigner-algorithm as this transitive dependency does
+      // not exist on mavencentral and hence cannot be found in the test environment
+      sql(s"ADD JAR ivy://org.apache.hive.hcatalog:hive-hcatalog-core:$hiveVersion" +
+        "?exclude=org.pentaho:pentaho-aggdesigner-algorithm")
```
Comment (xkrogen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:1230]: This is odd. Any idea why `hive-hcatalog-core` has a dependency on a package that doesn't exist?

**Review by xkrogen — COMMENTED** [review]
Great finds @shardulm94 ! I'm glad we were able to catch this before it went out in a release (since #29966 missed the 3.1.1 release). Thanks a lot for doing this and PR #31591 to really polish up this feature!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala (hunk 1 above)
Comment (shardulm94) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:1230]: This is a very infamous dependency brought in by calcite. It does not exist on mavencentral but it does exist on conjars.org. Here is a Stack overflow asking about the same issue. https://stackoverflow.com/questions/32587769/dependency-resolution-issue-in-gradle

Comment (xkrogen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:1230]: Yikes.

**Comment by AngersZhuuuu** [comment]
Thanks for your double check, nice catch,  I miss some point when check hive's behavior.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @maropu and @dongjoon-hyun too FYI

**Comment by maropu** [comment]
Oh, I see. Nice catch.

> The Hive behavior is that transitive is true if not specified in the coordinate and false for invalid values. 

Just to check; all the released Hive versions having this feature have the same default behaviour?

File: core/src/main/scala/org/apache/spark/util/DependencyUtils.scala [hunk 2]
```diff
@@ -59,8 +59,8 @@ private[spark] object DependencyUtils extends Logging {
    * @param uri Ivy URI need to be downloaded.
    * @return Tuple value of parameter `transitive` and `exclude` value.
    *
-   *         1. transitive: whether to download dependency jar of Ivy URI, default value is false
-   *            and this parameter value is case-sensitive. Invalid value will be treat as false.
+   *         1. transitive: whether to download dependency jar of Ivy URI, default value is true
+   *            and this parameter value is case-insensitive. Invalid value will be treat as false.
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/DependencyUtils.scala:63]: Could you leave some comments like `this behaviour follows hive one...`?

File: docs/sql-ref-syntax-aux-resource-mgmt-add-jar.md [hunk 3]
```diff
@@ -36,7 +36,7 @@ ADD JAR file_name
     The name of the JAR file to be added. It could be either on a local file system or a distributed file system or an Ivy URI.
     Apache Ivy is a popular dependency manager focusing on flexibility and simplicity. Now we support two parameter in URI query string:
 
-     * transitive: whether to download dependent jars related to your ivy URL. It is case-sensitive and only take last one if multiple transitive parameters are specified.
+     * transitive: whether to download dependent jars related to your ivy URL. The parameter name is case-sensitive, parameter value is case-insensitive and only take last one if multiple transitive parameters are specified.
```
Comment (maropu) [inline docs/sql-ref-syntax-aux-resource-mgmt-add-jar.md:39]: Could you split the second statement into two parts? e.g.,
```
The parameter name is case-sensitive and the parameter value is case-insensitive. If multiple transitive parameters are specified, the last one wins.
```

**Comment by maropu** [comment]
I left some minor comments though, the proposal itself seems okay.

File: core/src/main/scala/org/apache/spark/util/DependencyUtils.scala (hunk 2 above)
Comment (shardulm94) [reply core/src/main/scala/org/apache/spark/util/DependencyUtils.scala:63]: Fixed

File: docs/sql-ref-syntax-aux-resource-mgmt-add-jar.md (hunk 3 above)
Comment (shardulm94) [reply docs/sql-ref-syntax-aux-resource-mgmt-add-jar.md]: Done.

**Comment by shardulm94** [comment]
> Just to check; all the released Hive versions having this feature have the same default behaviour?

Yes, the behaviour has been this way since it was added in [HIVE-9664](https://issues.apache.org/jira/browse/HIVE-9664).

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
If no one has more comments, I'll merge this in a few days.

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by HyukjinKwon** [comment]
Thanks @maropu for taking care of this!

**Comment by dongjoon-hyun** [comment]
Thank you, @shardulm94 and all. Like @xkrogen said, it's really nice to fix this at the early stage.

**Pull request closed without merging**